subcalc.1
.\" $Header: /usr/cvs/subcalc/subcalc.1,v 1.6 2003/08/24 07:41:01 modulus Exp $
.Dd March 17, 2016
.Dt SUBCALC 1
.Os
.Sh NAME
.Nm subcalc
.Nd Advanced IP/IPv6 subnet calculation and discovery
.Sh SYNOPSIS
.Nm 
.Op Ar address\ family
.Op Ar address
.Op Cm print
.Nm
.Op Ar address\ family
.Op Ar address
.Op Cm netmask
.Op Ar mask
.Op Cm print
.Nm
.Op Ar address\ family
.Op Ar address
.Op Cm mask
.Op Ar ciscomask
.Op Cm print
.Nm
.Op Ar address\ family
.Op Ar address
.Op Cm prefixlen
.Op Ar len
.Op Cm print
.Nm
.Op Ar address\ family
.Op Cm hosts
.Op Ar number
.Nm
.Op Cm int6
.Op Ar address
.Op Ar hostname
.Nm
.Op Cm arpa6
.Op Ar address
.Op Ar hostname
.Nm
.Op Cm stf
.Op Ar address\ family
.Op Ar address
.Sh DESCRIPTION
The
.Nm
utility is used for subnet calculation and IPv6 DNS PTR record generation.
.Nm
takes command line arguments in a similar format to ifconfig(8) so
the synopsis should be familiar to the user. Given an address
family, address and a netmask/prefix length,
.Nm
will calculate the number of hosts and address ranges of the
specific network. Specifying the ``print'' option to to the
end of the command line will result in every single network
address for the specified network being printed to stdout.
.Pp
.Nm
was designed for network engineers which setup fine
grained firewalls, packet filters, access control lists and
network subdivisions for both IP and IPv6 servers and networks.
.Pp
The following options are available:
.Bl -tag -width indent
.It Ar address_family
Specify the
address family
which affects interpretation of the remaining parameters.
Since
.Nm
can handle subnet calculations in different protocols
with different naming schemes, specifying the address family is recommended.
The address or protocol families currently
supported are
.Dq inet
and 
.Dq inet6 .
.It Ar address
For the
.Tn DARPA Ns -Internet
family,
the address is
an Internet address expressed in the Internet standard
.Dq dot notation .
.Pp
It is also possible to use the CIDR notation (also known as the
slash notation) to include the netmask.
That is, one can specify an address like
.Li 192.168.0.1/16 .
.Pp
For
.Dq inet6
family, it is also possible to specify the prefix length using the slash
notation, like
.Li ::1/128 .
See the
.Cm prefixlen
parameter below for more information.
.It Cm arpa6 Ar address
Generate an ARPA6 PTR record which the popular DNS server ``BIND''
will understand. In order for the components of the PTR record
to be complete, a hostname must be specified as well.
.It Cm hosts Ar number
Specify the number of nodes or addresses on a network and
.Nm
will calculate an appropriate network subdivision and bits. This
is used for netmask discovery. NOTE: currently
.Nm
calculates the subdivision for one network. In some cases
it may be more efficient to use more then one network subdivision.
.It Cm int6 Ar address
Generate an INT6 PTR record which the popular DNS server ``BIND''
will understand. In order for the components of the PTR record
to be complete, a hostname must be specified as well. 
.It Cm mask Ar ciscomask
Specify the netmask in the format expected on Cisco routers. That is
the inverted form of a netmask. For example, the netmask 255.255.255.0
would be represented as 0.0.0.255. This option is only applicable for
IPv4, as on Cisco routers, CIDR notation is used to specify IPv6 network
addresses.
.It Cm netmask Ar mask
Specify how much of the address to reserve for subdividing
networks into sub-networks.
The mask includes the network part of the local address
and the subnet part, which is taken from the host field of the address.
The mask can be specified as a single hexadecimal number
with a leading
.Ql 0x ,
or with a dot-notation Internet address.
The mask contains 1's for the bit positions in the 32-bit address
which are to be used for the network and subnet parts,
and 0's for the host part.
The mask should contain at least the standard network portion,
and the subnet field should be contiguous with the network
portion.
.Pp
The netmask can also be specified in CIDR notation after the address.
See the
.Ar address
option above for more information.
.It Cm prefixlen Ar len
Specify that
.Ar len
bits are reserved for subdividing networks into sub-networks.
The
.Ar len
must be an integer, and for syntactical reason it must be between 0 to 128.
It is almost always 64 under the current IPv6 assignment rule.
If the parameter is omitted, 64 is used.
.Pp
The prefix can also be specified using the slash notation after the address.
See the
.Ar address
option above for more information
.It Cm print
causes each address in the specified network to be printed
to stdout, one address per line.  Regular subnet calculation
is still performed, however each line of the standard
calculation output is prefixed by a ``;'' character. This
makes it easy for standard subnet calculations to be filtered
out, in the event that network lists need to be generated.
.It Cm stf
If the address family is inet, translate the argument as
an IPv4 address which will be converted into the arbitrary
IPv6 ``/48'' (6to4 rfc3056) network designation. If the family
is inet6, translate the argument as an IPv6 network or address
specification which will be converted into it's arbitrary
IPv4 parent address.
.Sh EXAMPLES
To calculate the network range, number of hosts, prefixlen
or CIDR and netmask for the 10.0.0.1/24 (255.255.255.0) network.
.Pp
.D1 % subcalc inet 10.0.0.1/24
.Pp
Anyone of the following will achieve the exact same thing:
.Pp
.D1 % subcalc inet 10.0.0.1 netmask 255.255.255.0
.D1 % subcalc inet 10.0.0.1 netmask 0xffffff00
.D1 % subcalc inet 10.0.0.1 prefixlen 24
.D1 % subcalc inet 10.0.0.1 mask 0.0.0.255
.Pp
To generate a list of nodes for the specified network
one could use anyone of the following methods:
.Pp
.D1 % subcalc inet 10.0.0.1/24 print
.D1 % subcalc inet 10.0.0.1 netmask 255.255.255.0 print
.D1 % subcalc inet 10.0.0.1 netmask 0xffffff00 print
.D1 % subcalc inet 10.0.0.1 prefixlen 24 print
.Pp
Arbitrarily, the same thing can be done for IPv6. To calculate
the network range, number of hosts, prefixlen etc for
the 3ffe:beef:13e1:4c92::cd90/48 network, one could use any
of the following:
.Pp
.D1 % subcalc inet6 3ffe:beef:13e1:4c92::cd90/48
.D1 % subcalc inet6 3ffe:beef:13e1:4c92::cd90 netmask ffff:ffff:ffff::
.D1 % subcalc inet6 3ffe:beef:13e1:4c92::cd90 prefixlen 48
.Pp
Likewise for printing each node of a network:
.Pp
.D1 % subcalc inet6 3ffe:beef:13e1:4c92::cd90/48 print
.D1 % subcalc inet6 3ffe:beef:13e1:4c92::cd90 prefixlen 48 print
.Pp
IP version 4 PTR records (records which map an address to a 
canonical name) contain a portion which has a reversed
version of the IP octet prefixing the in-addr.arpa. zone.
For example, an RFC 1035 PTR record for 10.0.0.1 would look
like:
.Pp
.D1 1.0.0.10.in-addr.arpa. IN PTR canonical.name.org.
.Pp
The first section of this PTR record is called the ``owner'':
.Pp
.D1 1.0.0.10.in-addr.arpa.
.Pp
For IP version 6, rather then reversing each octet, PTR
records require that each ``nibble'' or 4 bit address
subdivision be reversed. For example, the owner section
of a PTR record for the IPv6 address 3ffe:beef:13e1:4c92::cd90 would be:
.Pp
.D1 0.9.d.c.0.0.0.0.0.0.0.0.0.0.0.0.2.9.c.4.1.e.3.1.f.e.e.b.e.f.f.3.ip6.arpa.
.Pp
Creating PTR record for IP version 6 addresses can be tedious and more prone
to error.
.Nm
has the capability to automatically subdivide the IPv6 address into nibbles
and reverse them into a PTR record like format. An example on how this
might be achieved is:
.Pp
.D1 % subcalc arpa6 3ffe:beef:13e1:4c92::cd90 canonical.name.org.
.D1 % subcalc int6 3ffe:beef:13e1:4c92::cd90 canonical.name.org.
.Pp
If you wanted to calculate a subdivision for a network given
the number of nodes or hosts, you could do:
.Pp
.D1 % subcalc inet hosts 23
.Pp
The next highest network subdivision in this case would be
a ``/27'' which has 32 hosts. The netmask for this network
would be: 255.255.255.224.
.Pp
Given an IPv4 address, figure out the 6to4 IPv6 ``/48'' network
associated with it.
.Pp
.D1 % subcalc stf inet 204.55.55.1
.Pp
Given an IPv6 network or address, figure out what IPv4 address
represents the parent for the network.
.Pp
.D1 % subcalc stf inet6 2002:cc37:3701::/48
.D1 % subcalc stf inet6 2002:cc37:3701:bc38::4081
.Pp
.Sh AUTHOR
The
.Nm
utility and this man page was written by Christian S.J. Peron.
Send any bugs or patches to (csjp@sqrt.ca).
.Sh SEE ALSO
.Xr ifconfig 8 ,
.Xr inet 4 ,
.Xr netintro 4 , 
.Xr netstat 1 ,
.Xr stf 4



subcalc.c
/*-
 * Copyright (c) Christian S.J. Peron (csjp@sqrt.ca) 
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The names of the authors may not be used to endorse or promote
 *    products derived from this software without specific prior written
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 */
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include <unistd.h>
#include <assert.h>
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <err.h>

#include "subcalc.h"

static int dorange = 0;
static char *prog = "subcalc";

static int usage(void);
static int packadrinfo(int af, u_char *adrspace, const char *str);
static char *getipaddress(int af, u_char *adrspace);
static int pl2m[9] = { 0x00, 0x80, 0xc0, 0xe0, 0xf0, 0xf8, 0xfc, 0xfe, 0xff };
static char *invert_mask(int, void *, void *);

static int
mask_discover(char *number, int af)
{
	double lg, lg2, pwr;
	u_int i, sp;
	int mx;

	sp = atoi(number);
	if (sp == 0) {
		(void) fprintf(stderr, "invalid mask specification: %s\n",
		    number);
		exit(1);
	}
	mx = (af == AF_INET6) ? 128 : 32;
	for (i = 0; i <= mx; i++) {
		pwr = pow(2, i);
		if (pwr >= sp)
			break;
	}
	lg = log((double)sp) / log(2);
	lg2 = log(pwr) / log(2);
	printf("theoretical len: %.30f\n", (double)(mx - lg));
	printf("working len:     %.0f\n", (double)(mx - lg2));
	printf("hosts:           %.0f\n", pow(2, lg2));
	return (0);
}

struct in6_addr *
plen2mask(int n)
{
	static struct in6_addr ia;
	u_char *p;
	int i;

	memset(&ia, 0, sizeof(struct in6_addr));
	p = (u_char *)&ia;
	for (i = 0; i < 16; i++, p++, n -= 8) {
		if (n >= 8) {
			*p = 0xff;
			continue;
		}
		*p = pl2m[n];
		break;
	}
	return (&ia);
}

static int
setaddrmask(struct in6_addr *ip, struct in6_addr *mask, unsigned bits)
{
	int i;

	*mask = *(plen2mask(bits));
	for(i = 0; i < sizeof(*ip); i++)
		ip->s6_addr[i] &= mask->s6_addr[i];
	return (0);
}

static int	extractbits(int af, u_char *adrspace);

struct cmdargs {
	int af;
	char address[64];
	u_long bits;
};

static int
proccmdargs(int c, char *a[], struct cmdargs *p)
{
	static char **fp, *fields[10];
	struct in_addr tmp_mask;
	struct in6_addr a6;
	char *m, *r, *tmp;
	int g;
	union {
		struct in_addr mask;
		struct in6_addr mask6;
	} msk;
	union {
		u_int32_t a;
		u_char octet[4];
	} stf;

	if (strcmp(a[1], "stf") == 0) {
		if (&a[2][0] == NULL)
			errx(1, "must specify an address family");
		else if (strcmp(a[2], "inet") == 0) {
			if (&a[3][0] == 0)
				errx(1, "must specify ip address");
			packadrinfo(AF_INET, (u_char *)&stf.a, a[3]);
		} else if (strcmp(a[2], "inet6") == 0) {
			if (&a[3][0] == 0)
				errx(1, "must specify ip6 address or net");
			sscanf(a[3], "2002:%02x%02x:%02x%02x:",
			    (unsigned int *)&stf.octet[0], (unsigned int *)&stf.octet[1],
			    (unsigned int *)&stf.octet[2], (unsigned int *)&stf.octet[3]);
		}
		printf("6to4 network:        2002:%02x%02x:%02x%02x::/48\n",
		    stf.octet[0], stf.octet[1], stf.octet[2], stf.octet[3]);
		printf("ip version 4 parent: %s\n",
		    getipaddress(AF_INET, (u_char *)&stf.a));
		exit(0);
	}
	g = c - 1;
	if (strcmp(a[1], "int6") == 0) {
		if (&a[2][0] == NULL)
			errx(1, "must specify ip6 address.");
		if (&a[3][0] == NULL)
			errx(1, "must specify a hostname.");
		packadrinfo(AF_INET6, (u_char *)&a6, a[2]);
		for (g = 15; g >= 0; g--)
			printf("%x.%x.%s",
			    (a6.s6_addr8[g] & 0x0f),
			    (a6.s6_addr8[g] & 0xf0) >> 4,
			    (g == 0) ? "ip6.int.\tIN\tPTR\t" : "");
		printf("%s\n", a[3]);
		exit(0);
	}
	if (strcmp(a[1], "arpa6") == 0) {
		if (&a[2][0] == NULL)
			errx(1, "must specify ip6 address.");
		if (&a[3][0] == NULL)
			errx(1, "must specify a hostname.");
		packadrinfo(AF_INET6, (u_char *)&a6, a[2]);
		for (g = 15; g >= 0; g--)
			printf("%x.%x.%s",
			    (a6.s6_addr8[g] & 0x0f),
			    (a6.s6_addr8[g] & 0xf0) >> 4,
			    (g == 0) ? "ip6.arpa.\tIN\tPTR\t" : "");
		printf("%s\n", a[3]);
		exit(0);
	}
	if (strcmp(a[g], "print") == 0)
		dorange++;
	if (strcmp("inet", a[1]) == 0) {
		if (c < 3)
			usage();
		p->af = AF_INET;
		if (strcmp(a[2], "hosts") == 0) {
			if (&a[3][0] == NULL)
				usage();
			mask_discover(a[3], AF_INET);
			exit(0);
		}
		if (strchr(a[2], '/')) {
			tmp = &a[2][0];
			for (fp = fields; (*fp = strsep(&tmp, "/"))
				!= NULL;) {
				if (**fp != '\0')
					if (++fp >= &fields[2])
						break;
			}
			if (fields[0] == NULL || fields[1] == NULL) {
				(void) fprintf(stderr, "invalid address/cidr specification\n");
				exit(1);
			}
			memcpy(p->address, fields[0], sizeof(p->address));
			p->bits = atoi(fields[1]);
			return (0);
		} else if (c == 3)
			errx(1, "specify network bits or mask.");
		memcpy(p->address, a[2], sizeof(p->address));
		if (strcmp(a[3], "netmask") == 0) {
			if (c != 5 && !dorange)
				errx(1,"invalid words near netmask");
			m = &a[4][0];
			if (*m == '0' && *m + 1 == 'x')
				msk.mask.s_addr = strtoul(m, &r, 16);
			else
				msk.mask.s_addr = inet_addr(a[4]);
			p->bits = extractbits(AF_INET, (u_char *)&msk.mask);
		}
		if (strcmp(a[3], "mask") == 0) {
			if (c != 5 && !dorange)
				errx(1,"invalid words near mask");
			tmp_mask.s_addr = inet_addr(a[4]);
			(void) invert_mask(AF_INET, &tmp_mask,
			    &msk.mask.s_addr);
			p->bits = extractbits(AF_INET, (u_char *)&msk.mask);
		}
		if (strcmp(a[3], "prefixlen") == 0) {
			if (c != 5 && !dorange)
				errx(1,"invalid words near prefixlen");
			p->bits = atoi(a[4]);
		}
		return (0);
	}

	if (strcmp("inet6", a[1]) == 0) {
		if (c < 3)
			usage();
		p->af = AF_INET6;
		if (strcmp(a[2], "hosts") == 0) {
			if (&a[3][0] == NULL)
				usage();
			mask_discover(a[3], AF_INET6);
			exit(0);
		}
		if (strchr(a[2], '/')) {
			tmp = &a[2][0];
			for(fp = fields; (*fp = strsep(&tmp, "/"))
				!= NULL;) {
				if (**fp != '\0')
					if (++fp >= &fields[2])
						break;
			}
			if (fields[0] == NULL || fields[1] == NULL) {
				(void) fprintf(stderr, "invalid address/cidr specification\n");
				exit(1);
			}
			memcpy(p->address, fields[0], sizeof(p->address));
			p->bits = atoi(fields[1]);
			return (0);
		} else if (c == 3)
			errx(1, "specify network bits or mask.");
		memcpy(p->address, a[2], sizeof(p->address));
		if (strcmp(a[3], "prefixlen") == 0) {
			if (c != 5 && !dorange)
				errx(1,"invalid words near prefixlen");
			p->bits = atoi(a[4]);
		}
		if (strcmp(a[3], "netmask") == 0) {
			if (c != 5 && !dorange)
				errx(1,"invalid words near netmask");
			inet_pton(AF_INET6, a[4], (u_char *)&msk.mask6);
			p->bits = extractbits(AF_INET6, (u_char *)&msk.mask6);
		}
		return (0);
	}
	errx(1, "`%s' is an invalid address family.", a[1]);
	return (0);
}

static int
getb(u_char *field, unsigned pos)
{
	u_char mask;
	int i;

	mask = 0x80;
	for(i = 0; i < (pos % 8); i++)
		mask = (mask >> 1);
	return (((mask & field[(int)(pos / 8)]) == mask) ? 1 : 0);
}

static int
setb(u_char *field, unsigned pos, char state)
{
	u_char mask;
	int i;

	mask = 0x80;
	for(i = 0; i < (pos % 8); i++)
		 mask = (mask >> 1);
	if (state)
		field[pos / 8] |= mask;
	else
		field[pos / 8] &= ~mask;
	return (0);
}

static char *
invert_mask(int af, void *addr, void *conv)
{
	char buf[128];
	int i, bit;
	union {
		struct in6_addr *in6;
		struct in_addr *in;
	} adu;

	bzero(&buf[0], sizeof(buf));
	/*
	 * NB: This function was created to effectively invert the subnet mask
	 * to be used with Cisco's ACL specifications.  Last time I looked at
	 * the Cisco configuration for ACLs they used CIDR and not the
	 * masking method.
	 */
	assert(af == PF_INET);
	adu.in = (struct in_addr *)addr;
	for (i = IPWIDTH - 1; i >= 0; i--) {
		bit = getb((u_char *)&adu.in->s_addr, i);
		if (bit == 0)
			setb((u_char *)&adu.in->s_addr, i, 1);
		else
			setb((u_char *)&adu.in->s_addr, i, 0);
	}
	if (conv != NULL)
		(void) bcopy(&adu.in->s_addr, conv, sizeof(adu.in->s_addr));
	inet_ntop(af, &adu.in->s_addr, &buf[0], INET_ADDRSTRLEN);
	return (strdup(&buf[0]));
}

static int
unsetmask(int af, u_char *adrspace, unsigned b)
{
	int i;
	union {
		struct in6_addr *in6;
		struct in_addr *in;
	} adu;

	if (af == AF_INET6) {
		adu.in6 = (struct in6_addr *)adrspace;
		for(i = IPV6WIDTH-1; i >= (IPV6WIDTH - b); i--)
			setb((u_char *)&adu.in6->s6_addr8[0], i, 0);
	}
	if (af == AF_INET) {
		adu.in = (struct in_addr *)adrspace;
		for (i = IPWIDTH-1; i >= (IPWIDTH - b); i--)
			setb((u_char *)&adu.in->s_addr, i, 0);
	}
	return (0);
}

static int
extractbits(int af, u_char *adrspace)
{
	unsigned bits;
	int i;

	bits = 0;
	if (af == AF_INET6) {
		for(i = 0; i < IPV6WIDTH; i++)
			if (!getb((u_char *)adrspace, i))
				bits++;
		return (IPV6WIDTH - bits);
	}
	if (af == AF_INET) {
		for(i = 0; i < IPWIDTH; i++)
			if (!getb((u_char *)adrspace, i))
				bits++;
		return (IPWIDTH - bits);
	}
	return (-1);
}

static int
packadrinfo(int af, u_char *adrspace, const char *str)
{
	union {
		struct in6_addr *in6;
		struct in_addr *in;
	} adu;

	if (af == AF_INET6) {
		adu.in6 = (struct in6_addr *)adrspace;
		inet_pton(AF_INET6, str, adu.in6);
	}
	if (af == AF_INET) {
		adu.in = (struct in_addr *)adrspace;
		adu.in->s_addr = inet_addr(str);
		if (adu.in->s_addr == INADDR_NONE)
			errx(1, "invalid address specification");
	}
	return (0);
}

static int
setmask(int af, u_char *adrspace, unsigned b)
{
	int i;
	union {
		struct in6_addr *in6;
		struct in_addr *in;
	} adu;

	if (af == AF_INET6) {
		adu.in6 = (struct in6_addr *)adrspace;
		for (i = IPV6WIDTH-1; i >= (IPV6WIDTH - b); i--)
			setb((u_char *)&adu.in6->s6_addr8[0], i, 1);
	}
	if (af == AF_INET) {
		adu.in = (struct in_addr *)adrspace;
		for (i = IPWIDTH-1; i >= (IPWIDTH - b); i--)
			setb((u_char *)&adu.in->s_addr, i, 1);
	}
	return (0);
}

static char *
getipaddress(int af, u_char *adrspace)
{
	static char buf[64];

	bzero(&buf[0], sizeof(buf));
	inet_ntop(af, adrspace, buf, sizeof(buf));
	return (&buf[0]);
}

static int
usage(void)
{

	(void) fprintf(stderr,
	    "usage: %s [inet | inet6] address [netmask | mask ] mask <print>\n"
	    "       %s [inet | inet6] address [prefixlen] len <print>\n"
	    "       %s [inet | inet6] hosts value\n"
	    "       %s [int6 | arpa6] address hostname\n"
	    "       %s stf [inet | inet6 ] address\n",
	    prog, prog, prog, prog, prog);
	exit(1);
	/* NOT REACHED */
}

int
main(int argc, char *argv [])
{
	struct in6_addr adr6, adr62, ip6, ip6mask;
	u_int destmask, valmask;
	struct in_addr adr, adr2;
	char buf[64], *cmask;
	struct cmdargs cd;
	u_char *aaa;
	double p;
	int b, x;

	if (argc == 1)
		usage();
	proccmdargs(argc, argv, &cd);
	if (cd.af == AF_INET6) {
		SETADR6(adr6);
		SETADR6(adr62);
		SETADR6(ip6);
		setaddrmask(&ip6, &ip6mask, cd.bits);
		b = IPV6WIDTH - cd.bits;
		unsetmask(AF_INET6, (u_char *)&adr6, b);
		printf("%srange:       %s > ",
		    (dorange ? "; " : ""), getipaddress(AF_INET6,
		    (u_char *)&adr6));
		setmask(AF_INET6, (u_char *)&adr62, b);
		printf("%s\n", getipaddress(AF_INET6, (u_char *)&adr62));
		p = pow(2, (double)b);
		printf("%shosts:       %.0f\n",
		    (dorange ? "; " : ""), p);
		printf("%sprefixlen:   %lu\n",
		    (dorange ? "; " : ""), cd.bits);
		printf("%smask:        %s\n",
		    (dorange ? "; " : ""), inet_ntop(AF_INET6, &ip6mask, buf,
		    sizeof(buf)));
		if (dorange == 0)
			return (0);
		for(;;) {
			x = 15;
			if (MASKEQUAL(&adr6, &ip6mask, &ip6))
				printf("%s\n", getipaddress(AF_INET6, 
				    (u_char *)&adr6));
			else
				break;
			while (x >= 0 && (++adr6.s6_addr8[x] & 0xff) == 0)
				x--;
		}
	}
	if (cd.af == AF_INET) {
		packadrinfo(AF_INET, (u_char *)&adr, cd.address);
		packadrinfo(AF_INET, (u_char *)&adr2, cd.address);
		b = IPWIDTH - cd.bits;
		unsetmask(AF_INET, (u_char *)&adr, b);
		printf("%srange:       %s > ",
		   (dorange ? "; " : ""),
		    getipaddress(AF_INET, (u_char *)&adr));
		setmask(AF_INET, (u_char *)&adr2, b);
		printf("%s\n", getipaddress(AF_INET, (u_char *)&adr2));
		printf("%srange b10:   %u > %u\n",
		    (dorange ? "; " : ""), htonl(adr.s_addr),
		    htonl(adr2.s_addr));
		printf("%srange b16:   0x%x > 0x%x\n",
		    (dorange ? "; " : ""), htonl(adr.s_addr),
		    htonl(adr2.s_addr));
		adr2.s_addr = htonl(~0 << (32 - cd.bits));
		p = pow(2, (double)b);
		printf("%shosts:       %.0f\n",
		    (dorange ? "; " : ""), p);
		printf("%sprefixlen:   %lu\n",
		    (dorange ? "; " : ""), cd.bits);
		printf("%snetmask:     %s\n",
		    (dorange ? "; " : ""), getipaddress(AF_INET,
		    (u_char *)&adr2));
		cmask = invert_mask(AF_INET, &adr2, NULL);
		printf("%smask:        %s\n",
		    (dorange ? "; " : ""), cmask);
		if (dorange == 0)
			return (0);
		destmask = 1 << b;
		valmask = 0;
		while (valmask != destmask) {
			x = 3;
			aaa = (u_char *)&adr.s_addr;
			printf("%s\n", getipaddress(AF_INET, 
				(u_char *)&adr));
			while (x >= 0 && ((++aaa[x] & 0xff) == 0))
				x--;
			valmask++;
		}
	}
	return (0);
}
